### PR TITLE
upcoming: [M3-7702] - Update Region Select for edge sites

### PR DESCRIPTION
**Pull request by hana-linode**
## Description 📝
Update our Region Select to support edge regions for Project Gecko

## Changes  🔄
- New `site_type` property to the Region type
- New `gecko` feature flag
- New icon to indicate an Edge region and updates to support displaying it

## Preview 📷
image

### Prerequisites
(How to setup test environment)
- Turn on the MSW and Gecko feature flag in DevTools 

### Verification steps
(How to verify changes)
- Go to `/linodes/create` and open the Region Select dropdown. You should see an icon next to the `Gecko Edge Test` region and also icon helper text next to the region select.
- You should _not_ see the icon or icon helper text for Marketplace, Cloning, or Images

```
yarn test RegionSelect
yarn test RegionSelect.utils
```

## As an Author I have considered 🤔

*Check all that apply*

- [x] 👀 Doing a self review
- [ ] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [x] 🤏 Splitting feature into small PRs
- [x] ➕ Adding a changeset
- [x] 🧪 Providing/Improving test coverage
- [x] 🔐 Removing all sensitive information from the code and PR description
- [x] 🚩 Using a feature flag to protect the release
- [ ] 👣 Providing comprehensive reproduction steps
- [ ] 📑 Providing or updating our documentation
- [ ] 🕛 Scheduling a pair reviewing session
- [x] 📱 Providing mobile support
- [ ] ♿  Providing accessibility support

### CODE REVIEW

File: packages/api-v4/src/regions/types.ts [hunk 1]
```diff
@@ -22,13 +22,16 @@ export interface DNSResolvers {
...
+export type RegionSite = 'core' | 'edge';
+
 export interface Region {
   id: string;
   label: string;
   country: Country;
   capabilities: Capabilities[];
   status: RegionStatus;
   resolvers: DNSResolvers;
+  site_type: RegionSite;
```
Comment (bnussman-akamai) [inline packages/api-v4/src/regions/types.ts:34]: Is there a good reason they chose `site_type` instead of just `type`? A region is basically a "site", so this feels a bit redundant to me.

Comment (hana-linode) [reply packages/api-v4/src/regions/types.ts:34]: Carissa says the word `type` is reserved by a lot of languages so they try not to use that word though they haven't been great at enforcing it

Comment (bnussman-akamai) [reply packages/api-v4/src/regions/types.ts:34]: Understood! Thank you for checking on that!

File: packages/manager/src/components/Autocomplete/Autocomplete.tsx [hunk 2]
```diff
@@ -112,6 +112,12 @@ export const Autocomplete = <
           InputProps={{
             ...params.InputProps,
             ...textFieldProps?.InputProps,
+            endAdornment: (
+              <React.Fragment>
+                {textFieldProps?.InputProps?.endAdornment}
+                {params.InputProps.endAdornment}
+              </React.Fragment>
+            ),
```
Comment (hana-linode) [inline packages/manager/src/components/Autocomplete/Autocomplete.tsx:121]: Add support for displaying end adornments and the select dropdown arrow at the same time

image

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx [hunk 3]
```diff
@@ -38,6 +47,10 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
     width,
   } = props;
 
+  const flags = useFlags();
+  const location = useLocation();
+  const createType = getQueryParamFromQueryString(location.search, 'type');
```
Comment (bnussman-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.tsx:52]: It feels a bit odd that RegionSelect checks for Linode Create page query params. 

Maybe we should make all of this edge site stuff controllable via props so that RegionSelect remains generic?

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx [hunk 4]
```diff
@@ -132,6 +179,15 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
         placeholder="Select a Region"
         value={selectedRegion}
       />
+      {showGeckoHelperText && ( // @TODO Gecko MVP: Add docs link
+        <StyledEdgeBox>
+          <EdgeServer />
+          <Typography sx={{ textWrap: 'nowrap' }}>
```
Comment (mjac0bs) [inline packages/manager/src/components/RegionSelect/RegionSelect.tsx:185]: Can we have this text `alignSelf` to `center`? It's just slightly too high right now:

This branch:
Screenshot 2024-02-20 at 1 05 26 PM

Centered:
Screenshot 2024-02-20 at 1 06 23 PM

File: packages/manager/src/featureFlags.ts [hunk 5]
```diff
@@ -47,6 +47,7 @@ export interface Flags {
   databaseScaleUp: boolean;
   databases: boolean;
   firewallNodebalancer: boolean;
+  gecko: boolean;
```
Comment (mjac0bs) [inline packages/manager/src/featureFlags.ts:50]: This PR slips the new feature flag in. Can we call that out in the PR description too, just for completeness?

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx [hunk 6]
```diff
@@ -132,6 +179,15 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
         placeholder="Select a Region"
         value={selectedRegion}
       />
+      {showGeckoHelperText && ( // @TODO Gecko MVP: Add docs link
+        <StyledEdgeBox>
```
Comment (mjac0bs) [inline packages/manager/src/components/RegionSelect/RegionSelect.tsx:183]: Can we fix this little gap on tablet view on down? 

Screenshot 2024-02-20 at 1 03 55 PM

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx [hunk 7]
```diff
@@ -69,14 +82,32 @@ export const RegionSelect = React.memo((props: RegionSelectProps) => {
     }
   }, [selectedId]);
 
+  // Hide edge sites from Marketplace, Cloning, and Images
+  const unsupportedEdgeEntities = [
+    'Clone Linode',
+    'Images',
+    'One-Click',
+  ].includes(createType);
```
Comment (bnussman-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.tsx:90]: I wonder if it will bother people that we won't allow users to deploy Marketplace apps to edge regions in Cloud Manager.

**Review by mjac0bs — COMMENTED** [review]
Confirmed that:
- I see the icon and helper text for the mock edge site in the Linodes region dropdown.
- I do not see the icon or helper text for the region dropdown in other places it is used.
- Light and dark mode color looks good.

Any test coverage to come for edge sites in the region dropdown?

I left just a few minor punctuation and styling suggestions.

Edit: Also seeing the following unit test failure locally, so [943b272](https://github.com/linode/manager/pull/10194/commits/943b272c9831759c46fcee4e2d84fefd5d9d9495) might not have addressed the issue completely. 
image

File: packages/manager/src/components/RegionSelect/RegionSelect.tsx (hunk 7 above)
Comment (hana-linode) [reply packages/manager/src/components/RegionSelect/RegionSelect.tsx]: We won't allow it for now but may in the future. The Gecko team needs time to test and curate a Marketplace for edge regions.

**Comment by hana-linode** [comment]
@mjac0bs @bnussman-akamai great feedback, addressed!

**Review by mjac0bs — APPROVED** [review]
Thanks for addressing the feedback -- confirmed the above styling issues look good now and thanks for test coverage!

Do we have a plan for accessibility of the end adornment in the region select via aria? Using voice over, that icon is not read out and I don't think the user is able to view the tooltip, so they would not be able to distinguish edge sites (other than by name).

Approving pending that is addressed. 

Also, I'm not sure what's going on with CI for this PR. It failed initially, so I restarted the PR in jenkins, and now it's showing e2es passing but not sure why the rest of CI is still pending. Unit tests pass locally.

File: packages/manager/src/featureFlags.ts (hunk 5 above)
Comment (abailly-akamai) [reply packages/manager/src/featureFlags.ts:50]: You may want to consider using a object flag as we discussed recently. Something tells me gecko is going to have more than one parameter we will want to control individually. CC @jaalah-akamai

File: packages/manager/src/components/SelectRegionPanel/SelectRegionPanel.tsx [hunk 8]
```diff
@@ -72,6 +74,24 @@ export const SelectRegionPanel = (props: SelectRegionPanelProps) => {
...
+
+  const showGeckoHelperText = Boolean(
+    geckoEnabled &&
+      currentCapability &&
+      regions.find(
+        (region) =>
+          region.site_type === 'edge' &&
+          region.capabilities.includes(currentCapability)
+      )
+  );
```
Comment (abailly-akamai) [inline packages/manager/src/components/SelectRegionPanel/SelectRegionPanel.tsx:93]: I'd love to see a `isGeckoEnabled` util to which you pass your entities (`useFlags` can be invoked in the util) and tested as such. In doing so, we need to make those entities type safe

File: packages/manager/src/featureFlags.ts (hunk 5 above)
Comment (hana-linode) [reply packages/manager/src/featureFlags.ts:50]: @abailly-akamai I'll look into using an object flag in a separate PR

File: packages/manager/src/components/RegionSelect/RegionSelect.styles.ts [hunk 9]
```diff
@@ -20,8 +20,51 @@ export const StyledAutocompleteContainer = styled(Box, {
...
+      height: 21,
+      marginLeft: 8,
+      marginRight: 8,
+      width: 24,
+    },
+    alignSelf: 'end',
+    color: 'inherit',
+    display: 'flex',
+    marginLeft: 8,
+    padding: '8px 0',
```
Comment (abailly-akamai) [inline packages/manager/src/components/RegionSelect/RegionSelect.styles.ts:57]: nitpick (no need to address but good to keep in mind for the future) - while sizing an svg with hard coded values is totally fine, we can use theme spacing units for, well, spacing, especially for values we know to be 8 or multiple of 8. Who knows if the compact theme will ever make a comeback :D

**Review by abailly-akamai — APPROVED** [review]
Thanks for the changes! - second pass looks great - 🚢 ✅
